service/history/nDCTransactionMgr.go

```diff
@@ -295,3 +318,54 @@ func (r *nDCTransactionMgrImpl) loadNDCWorkflow(
 	}
 	return newNDCWorkflow(ctx, r.domainCache, r.clusterMetadata, context, msBuilder, release), nil
 }
+
+func (r *nDCTransactionMgrImpl) reapplyEvents(
+	ctx ctx.Context,
+	reapplyEvents *persistence.WorkflowEvents,
+) error {
+
+	hasReappliableEvent := false
+	for _, event := range reapplyEvents.Events {
+		switch event.GetEventType() {
+		case shared.EventTypeWorkflowExecutionSignaled:
+			hasReappliableEvent = true
+			break
+		default:
+		}
+	}
+	// there is no event to reapply
+	if !hasReappliableEvent {
+		return nil
+	}
+
+	domainID := reapplyEvents.DomainID
+	execution := &shared.WorkflowExecution{
+		WorkflowId: common.StringPtr(reapplyEvents.WorkflowID),
+		RunId:      common.StringPtr(reapplyEvents.RunID),
+	}
+	domainEntry, err := r.domainCache.GetDomainByID(domainID)
+	if err != nil {
+		return err
+	}
+
+	// The active cluster of the domain is the same as current cluster.
+	// Use the history from the same cluster to reapply events
+	activeCluster := domainEntry.GetReplicationConfig().ActiveClusterName
+	if activeCluster == r.shard.GetClusterMetadata().GetCurrentClusterName() {
+
+		return r.clientBean.GetHistoryClient().ReapplyEvents(ctx, &h.ReapplyEventsRequest{
```

Comment:
why going through RPC layer again? 
there is a `shard.GetEngine()` method which will bypass RPC layer

Reply:
done